### DISCUSSION TITLE
docs: Cascade persist from the inverse side

### DISCUSSION THREAD

**Discussion by univerze** [mikro-orm/mikro-orm#2595] [Q&A]
Hello, I have been browsing the docs and examples but could not make my relation persisted from the inverse side.

Based on the docs:

"Bidirectional are defined on both sides, while one is owning side (where references are store), marked by inversedBy attribute pointing to the inverse side. On the inversed side we define it with mappedBy attribute pointing back to the owner. When modeling bidirectional relationship, you can also omit the inversedBy attribute, defining mappedBy on the inverse side is enough as it will be auto-wired".

I'm using NestJS, here are my entities:
```
@Entity({ tableName: 'users' })
export class UserEntity {
  @PrimaryKey()
  id: string = v4();

  @ManyToMany({
    entity: () => GroupEntity,
    pivotTable: 'users_groups',
    joinColumn: 'user_id',
    inverseJoinColumn: 'group_id',
  })
  groups: Collection<GroupEntity> = new Collection<GroupEntity>(this);
}
```

```
@Entity({ tableName: 'groups' })
export class GroupEntity {
  @PrimaryKey()
  id: string = v4();

  @ManyToMany({
    entity: () => UserEntity,
    pivotTable: 'users_groups',
    joinColumn: 'group_id',
    inverseJoinColumn: 'user_id',
    mappedBy: user => user.groups
  })
  users: Collection<UserEntity> = new Collection<UserEntity>(this);
```

Persisting group relations from the user side works for the update, removes old entries from 'users_groups' table and adds the new relations coming from the dto payload:
```
    const user = await this.repo.findOneOrFail({ id });
    const groups = await this.groupRepo.find({ id: { $in: dto.groups } });
    user.groups.set(groups);
    await this.repo.persistAndFlush(user);
```

I expected the cascade persist work from the inverse side too, since the docs say it:
"When persisting or removing entity, all your references are by default cascade persisted. This means that by persisting any entity, ORM will automatically persist all of its associations."
also
"Collections on both sides have to be initialized, otherwise propagation won't work.
Although this propagation works also for M:N inverse side, you should always use owning side to manipulate the collection."

So this should work?
```
    const group = await this.repo.findOneOrFail({ id }, ['users']);
    const users = await this.userRepo.find({ id: { $in: dto.users } });
    group.users.set(users);
    await this.repo.persistAndFlush(group);
```
When I run above, it does not trigger any insert or update queries on persisting.

What am I doing wrong?

**Comment by univerze** [comment]
UPDATE:
After adding the populate to the users, cascade persist seems to be working now:
```
    const group = await this.repo.findOneOrFail({ id }, ['users']);
    const users = await this.userRepo.find({ id: { $in: dto.users }, populate: ['groups'] });
    group.users.set(users);
    await this.repo.persistAndFlush(group);
```

However on update it still doesn't cascade when removing users collection no matter how I set the cascade flags:
```
    const group = await this.repo.findOneOrFail({ id }, ['users']);
    group.deletedAt = new Date();
    group.users.removeAll();
    await this.repo.persistAndFlush(group);
```

I'm using update because we want to soft delete items rather than remove them from the tables.

**Comment by B4nan** [comment]
https://mikro-orm.io/docs/cascading/#cascade-persist

> When cascade persisting collections, keep in mind only fully initialized collections will be cascade persisted.

In other words, you always need to have the owning side initialized, as that is where the diffing will happen - inverse sides only propagate the changes to owners.

> However on update it still doesn't cascade when removing users collection no matter how I set the cascade flags:

We are talking about cascade persist, that has nothing to do with removing anything. Persist means making the EM aware of that entity, so it can be synchronized with the database state when you call flush, it will never do anything else than insert or update query (and you dont need it for update queries, as managed entities are persisted automatically). 

You are confusing multiple things - cascading (persist or remove - two different things) and propagation (changing bidirectional relation on one side will propagate the change to the other). Cascade persist is enabled by default, and propagation also works by default. Regarding `removeAll()` and propagation, there were recent fixes, I believe in v5 only, so you might want to check how it behaves there.

If you want more help, try to prepare a reproduction - minimal but complete, check the existing issue tests for inspiration. 

https://github.com/mikro-orm/mikro-orm/tree/master/tests/issues

If you are starting with a new project or just trying out things, use v5. It is nearly ready and there are many additional bugfixes that can't be backported to v4.
